### PR TITLE
Fix typing in translations

### PR DESCRIPTION
**Pull request by dstansby**
# Fixes/Closes
More partial work towards https://github.com/napari/napari/issues/2751

# Description
This fixes typing errors in `napari.utils.translations`.

# References

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

- [x] existing test suite passes
- [x] mypy passes

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: napari/utils/translations.py [hunk 1]
```diff
@@ -673,9 +686,8 @@ def _load_language(
     str
         The language locale set by napari.
     """
-    default_config_path = Path(default_config_path)
-    if default_config_path.exists():
-        with open(default_config_path) as fh:
+    if (config_path := Path(default_config_path)).exists():
+        with open(config_path) as fh:
```
Comment (Czaki) [inline napari/utils/translations.py]:
```suggestion
        with config_path.open() as fh:
```

File: napari/utils/translations.py [hunk 2]
```diff
@@ -50,11 +50,19 @@ def _get_display_name(
...
-        dislay_name = loc.get_display_name(display_locale).capitalize()
+        display_name = loc.get_display_name(display_locale)
+        if display_name is not None:
+            display_name = display_name.capitalize()
+        else:
+            # to allow raising error inside try: block
+            raise RuntimeError(
+                f"Could not find {display_locale}"
+            )  # ruff: noqa: TRY301
+
```
Comment (Czaki) [inline napari/utils/translations.py]: maybe everything after import should be moved to else clause. 

Or just use multiple return pattern?

This code is not readable to me.

File: napari/utils/translations.py [hunk 3]
```diff
@@ -210,11 +218,11 @@ def __new__(
 
     def __init__(
         self,
-        domain: Optional[str] = None,
+        domain: str,
         msgctxt: Optional[str] = None,
         msgid: Optional[str] = None,
         msgid_plural: Optional[str] = None,
-        n: Optional[str] = None,
+        n: Optional[int] = None,
```
Comment (Czaki) [inline napari/utils/translations.py:223]: What is `n`? If you investigate this code then maybe you could add docstrings?

File: napari/utils/translations.py (hunk 2 above)
Comment (dstansby) [reply napari/utils/translations.py]: Thanks - agree this wasn't the best! How about now? I've used a `try: ... except: ... else: ...` pattern to make the code mroe readable, and remove the need for the ruff noqa.

File: napari/utils/translations.py (hunk 3 above)
Comment (dstansby) [reply napari/utils/translations.py:223]: I'm not sure exactly, but this typing fix is definitely correct. I'd advocate for docstrings being beyond scope of this PR.

File: napari/utils/translations.py [hunk 4]
```diff
@@ -274,33 +277,34 @@ def translation(self) -> str:
...
+                    self._n,
+                )
+            else:
+                translation = gettext.dnpgettext(
+                    self._domain,
+                    self._msgctxt,
+                    self._msgid,
+                    self._msgid_plural,
+                    self._n,
+                )
```
Comment (Czaki) [inline napari/utils/translations.py]: These control flows allow having `translation` undefined. Maybe last else should be dedent

File: napari/utils/translations.py (hunk 3 above)
Comment (Czaki) [reply napari/utils/translations.py:223]: The docstring is out of scope. But I do not understand this code. some comments may help. Or this PR may need to wait longer for some core-dev who is familiar with translations.

Comment (dstansby) [reply napari/utils/translations.py:223]: I added a signpost to https://docs.python.org/3/library/gettext.html in the docstring, which is where these arguments are explained.

File: napari/utils/translations.py (hunk 4 above)
Comment (dstansby) [reply napari/utils/translations.py]: I've tidied up the logic here to add an `else:` statement at the end, and hopefully make it a bit more readable.

File: napari/utils/translations.py [hunk 5]
```diff
@@ -386,33 +400,27 @@ def _dnpgettext(
...
+                    n,
+                )
+            else:
+                translation = gettext.dnpgettext(
+                    self._domain,
+                    msgctxt,
+                    msgid,
+                    msgid_plural,
+                    n,
                 )
```
Comment (Czaki) [inline napari/utils/translations.py]: Does this else should be dedent?

File: napari/utils/translations.py (hunk 3 above)
Comment (Carreau) [reply napari/utils/translations.py:223]: `n` is the number of items, some language have pluralizations that depends on the number of things.

https://lingohub.com/blog/2019/02/pluralization

Comment (Carreau) [reply napari/utils/translations.py:223]: `n` is the number of items, some language have pluralizations that depends on the number of things.

https://lingohub.com/blog/2019/02/pluralization

File: napari/utils/translations.py (hunk 5 above)
Comment (dstansby) [reply napari/utils/translations.py]: Thanks for catching, I've refactored this to simplify the logic, make it match the docstring, and have an `else:` at the end.

**Review by GenevieveBuckley — APPROVED** [review]
This looks ok to me.

And I know this wasn't something you originally wrote, but can we address Grzegorz's concern by changing the variable name `n` to something more descriptive and/or leaving a detailed comment when it is first defined?

**Comment by dstansby** [comment]
> @dstansby So if you wound source of the name `n`, then could you pot this information in docstring? To avoid such problems in future.

I've put a signpost at https://github.com/napari/napari/pull/5900/files#diff-7fb6d118bffa1de32a621be943d23098b0367637c71bd3923a8772c0cc09d03aR169-R170 - is there somewhere else you'd like this signpost to go?

**Review by jni — APPROVED** [review]

**Pull request merged**